### PR TITLE
fat_fs: Add reentrant zephyr support

### PR DESCRIPTION
**Pull request by Ox11**
This commit enables zephyr to configure the FatFs FF_FS_REENTRANT
option and support fs actions from multiple threads.
CONFIG_FS_FATFS_REENTRANT enables the option and provides zephyr
mutex wrappers.

Signed-off-by: Nicola Ochsenbein <Nicola.Ochsenbein@husqvarnagroup.com>

### CODE REVIEW

**Comment by de-nordic** [comment]
Missing west.yml manifest update to include change in external module https://github.com/zephyrproject-rtos/fatfs/pull/13/files:
https://docs.zephyrproject.org/latest/develop/modules.html#changes-to-existing-module

File: subsys/fs/Kconfig.fatfs [hunk 1]
```diff
@@ -155,6 +155,11 @@ config FS_FATFS_WINDOW_ALIGNMENT
 	  that, in worst scenario, value provided here may cause FATFS
 	  structure to have size of twice the value
 
+config FS_FATFS_REENTRANT
```
Comment (de-nordic) [inline subsys/fs/Kconfig.fatfs:158]: This depends on posix subsystem.

**Comment by de-nordic** [comment]
Is it possible to provide test scenarios, in Zephyr, to prove this works correctly?

**Comment by Ox11** [comment]
@de-nordic @carlescufi PR is updated. Could you comment?

**Comment by de-nordic** [comment]
> @Ox11 thanks for the PR! After #51766 got merged this needs rework, can you please make the changes necessary? cc @de-nordic

There is also need to merge this https://github.com/zephyrproject-rtos/fatfs/pull/15 and west update here https://github.com/zephyrproject-rtos/zephyr/pull/52505, as the new FAT FS version brings fixes to reentrant code and also changes how synchronization primitives are used, so it directly affects this PR.

**Review by de-nordic — CHANGES_REQUESTED** [review]
FIx compliance.
Separate commit for code and tests.
Missing commit with release notes update.

**Comment by Ox11** [comment]
@de-nordic I cant find the results of the compliance checks anymore. Is this because i force pushed the new changes out of habbit?
If so, could you please re-trigger them or help me find the old results :grimacing: ?

File: tests/subsys/fs/fat_fs_api/src/test_fat_file_reentrant.c [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+			NULL, NULL, NULL,
+			K_PRIO_PREEMPT(0), 0, K_NO_WAIT);
+
+	/* Make sure thread was able to lock mutex */
+	k_sem_take(&run_non_thread_sem, K_FOREVER);
+
+	/* File open should wait here, as the fs is locked. Therefore, automatic switch back to thread */
+	TC_PRINT("Open file\n");
+	res = fs_open(&filep, TEST_FILE, FS_O_CREATE | FS_O_RDWR);
+	zassert_false(res, "Err: File could not be opened [%d]\n", res);
```
Comment (de-nordic) [inline tests/subsys/fs/fat_fs_api/src/test_fat_file_reentrant.c:42]: `zasset_false` expects boolean; you probably would like to use `zassert_ok` here.

File: tests/subsys/fs/fat_fs_api/prj_reentrant.conf [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+CONFIG_FAT_FILESYSTEM_ELM=y
+CONFIG_DISK_DRIVER_FLASH=y
+CONFIG_SPI=y
+CONFIG_GPIO=y
+CONFIG_ZTEST=y
+CONFIG_ZTEST_NEW_API=y
+CONFIG_FLASH=y
+CONFIG_FLASH_MAP=y
+CONFIG_FS_FATFS_REENTRANT=y
+CONFIG_MULTITHREADING=y
```
Comment (de-nordic) [inline tests/subsys/fs/fat_fs_api/prj_reentrant.conf:13]: You do not have to make entire prj file to add these. The prj differs by two configs from the non reentrant one, so you can use `extra_config` option of testcases.yaml to run with original prj.conf.

**Review by de-nordic — CHANGES_REQUESTED** [review]
1) fix compliance;
2) add yourself to ~~CODEOWNERS~~ MAINTAINERS for the zfs_ffsystem.c you have just added;
3) Add release notes that mentions adding reenatrans support for FATFS and addition of the Kconfig option CONFIG_FS_FATFS_REENTRANT.

**Comment by Ox11** [comment]
> 1. fix compliance;
> 2. add yourself to ~CODEOWNERS~ MAINTAINERS for the zfs_ffsystem.c you have just added;
> 3. Add release notes that mentions adding reenatrans support for FATFS and addition of the Kconfig option CONFIG_FS_FATFS_REENTRANT.

Compliance is fixed, I have added myself to the maintainers and added a release notes comment.

File: tests/subsys/fs/fat_fs_api/src/test_fat_file_reentrant.c [hunk 4]
```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright (c) 2022 Husqvarna AB
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "test_fat.h"
+
+#if IS_ENABLED(CONFIG_FS_FATFS_REENTRANT)
```
Comment (de-nordic) [inline tests/subsys/fs/fat_fs_api/src/test_fat_file_reentrant.c:9]: Just use #ifdef. IS_ENABLED is for hacking `if` to work with `CONFIG_`s

File: tests/subsys/fs/fat_fs_api/src/test_fat.h [hunk 5]
```diff
@@ -1,13 +1,15 @@
 /*
  * Copyright (c) 2016 Intel Corporation.
  * Copyright (c) 2020 Nordic Semiconductor ASA
+ * Copyright (c) 2022 Husqvarna AB
```
Comment (de-nordic) [inline tests/subsys/fs/fat_fs_api/src/test_fat.h:4]: It is 23 now, took us long to review.

File: doc/releases/release-notes-3.3.rst [hunk 6]
```diff
@@ -401,6 +401,7 @@ Libraries / Subsystems
   * Added new API call `fs_mkfs`.
   * Added new sample `samples/subsys/fs/format`.
   * FAT FS driver has been updated to version 0.15 w/patch1.
+  * Added FAT FS reentrant Kconfig :kconfig:option:`CONFIG_FS_FATFS_REENTRANT`.
```
Comment (de-nordic) [inline doc/releases/release-notes-3.3.rst:404]: The Kconfig is not reentrant. Rephrase this to be more clear that the Koconfig enables reentrant support in FAT FS.

File: subsys/fs/Kconfig.fatfs [hunk 7]
```diff
@@ -211,6 +212,11 @@ config FS_FATFS_WINDOW_ALIGNMENT
 	  that, in worst scenario, value provided here may cause FATFS
 	  structure to have size of twice the value.
 
+config FS_FATFS_REENTRANT
+	bool "FatFs reentrant"
+	help
+	  Enable the re-entrancy (thread safe) option of the FatFs
+
```
Comment (de-nordic) [inline subsys/fs/Kconfig.fatfs:219]: Add more info on costs: for example that there will be mutex objects, one per volume. This is important because this takes some RAM, so people need to know.

**Review by de-nordic — CHANGES_REQUESTED** [review]
Minor comments.

File: tests/subsys/fs/fat_fs_api/src/test_fat.h (hunk 5 above)
Comment (Ox11) [reply tests/subsys/fs/fat_fs_api/src/test_fat.h:4]: Shall I also update the git commit date?

File: modules/fatfs/zfs_ffsystem.c [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+/* Free a memory block */
+void ff_memfree(void *mblock)
+{
+	k_free(mblock);
+}
+#endif /* FF_USE_LFN == 3 */
+
+#if FF_FS_REENTRANT	/* Mutual exclusion */
+static struct k_mutex Mutex[FF_VOLUMES + 1]; /* Table of Zephyr mutex */
```
Comment (de-nordic) [inline modules/fatfs/zfs_ffsystem.c:30]: Where the + 1 comes from?

Comment (Ox11) [reply modules/fatfs/zfs_ffsystem.c]: The + 1 is for the system mutex. Same as in the template file by ChaN. I added a comment about it.

**Review by nordicjm — CHANGES_REQUESTED** [review]

File: modules/fatfs/CMakeLists.txt [hunk 9]
```diff
@@ -18,7 +18,10 @@ if(CONFIG_FAT_FILESYSTEM_ELM)
 
   zephyr_library_sources_ifdef(CONFIG_FS_FATFS_LFN
     ${ZEPHYR_FATFS_MODULE_DIR}/option/ffunicode.c
-    ${ZEPHYR_FATFS_MODULE_DIR}/option/ffsystem.c
+  )
+
+  zephyr_library_sources_ifdef((CONFIG_FS_FATFS_LFN OR CONFIG_FS_FATFS_REENTRANT)
+    zfs_ffsystem.c
```
Comment (nordicjm) [inline modules/fatfs/CMakeLists.txt:24]: This seems to be a wrong Kconfig order, if FS_FATFS_LFN mode requires reentrancy support, then it must pull FS_FATFS_REENTRANT in, either as a depends on or select, the this line is simplified

**Review by nordicjm — CHANGES_REQUESTED** [review]

File: modules/fatfs/CMakeLists.txt (hunk 9 above)
Comment (Ox11) [reply modules/fatfs/CMakeLists.txt:24]: They are not dependent on each other. Just both helper functions are implemented in the same file (same as in the example file provided by Elm Chan)
I did not get my local test running today, so I tried to test it directly on the pipeline :see_no_evil: 

Do you have a suggestion how I can include this zfs_ffsystem.c if either CONFIG_FS_FATFS_LFN or CONFIG_FS_FATFS_REENTRANT is set?

Comment (nordicjm) [reply modules/fatfs/CMakeLists.txt:24]: This is OK then, I misread how it works.

File: modules/fatfs/zfs_ffsystem.c [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+{
+	k_free(mblock);
+}
+#endif /* FF_USE_LFN == 3 */
+
+#if FF_FS_REENTRANT	/* Mutual exclusion */
+/* Table of Zephyr mutex. One for each volume and an extra one for the ff system.
+ * See also the template file used as reference. Link is available in the header of this file.
+ */
+static struct k_mutex Mutex[FF_VOLUMES + 1];
```
Comment (nordicjm) [inline modules/fatfs/zfs_ffsystem.c:33]: snake case - lowercase characters

**Review by nordicjm — CHANGES_REQUESTED** [review]

**Comment by Ox11** [comment]
@nordicjm @de-nordic Could you please re-trigger the pending workflows and re-review?

**Comment by de-nordic** [comment]
The window for enhancements to 3.3 has been closed now, so moving this to 3.4.

File: doc/releases/release-notes-3.3.rst [hunk 11]
```diff
@@ -510,6 +510,7 @@ Libraries / Subsystems
   * Added the option to disable CRC checking in :ref:`fcb_api` by enabling the
     Kconfig option :kconfig:option:`CONFIG_FCB_ALLOW_FIXED_ENDMARKER`
     and setting the `FCB_FLAGS_CRC_DISABLED` flag in the :c:struct:`fcb` struct.
+  * Added :kconfig:option:`CONFIG_FS_FATFS_REENTRANT` to enable the FAT FS reentrant option.
```
Comment (nordicjm) [inline doc/releases/release-notes-3.3.rst:513]: This will need to move to 3.4, but do that after 3.3 release is finished as the file will be created then

File: modules/fatfs/zephyr_fatfs_config.h [hunk 12]
```diff
@@ -66,6 +67,14 @@
 #define FF_FS_EXFAT		CONFIG_FS_FATFS_EXFAT
 #endif /* defined(CONFIG_FS_FATFS_EXFAT) */
 
+#if defined(CONFIG_FS_FATFS_REENTRANT)
+#undef FF_FS_REENTRANT
+#undef FF_FS_TIMEOUT
+#include <zephyr/kernel.h>
+#define FF_FS_REENTRANT		CONFIG_FS_FATFS_REENTRANT
+#define FF_FS_TIMEOUT		K_FOREVER
```
Comment (nordicjm) [inline modules/fatfs/zephyr_fatfs_config.h:75]: I'm not familiar with the internals of how this works, but let's assume that there is one partition with a log file and any other random file, if the log file is opened at startup and let's say it moves to a new file once an hour (or even once a day) but there's a way for files to be retrieved from this file system - what happens if a client requests the "other random file", does their file access part of their code (let's assume MCUmgr) then halt for **1 hour/1 day** until the mutex can be taken? If so, that seems rather inconvenient for all users of this, and unacceptable

**Review by nordicjm — CHANGES_REQUESTED** [review]

File: modules/fatfs/zephyr_fatfs_config.h (hunk 12 above)
Comment (Ox11) [reply modules/fatfs/zephyr_fatfs_config.h:75]: The reentrant option does not lock the media upon open. It only locks file access operations so that volume access is serialized. E.g. not two threads are trying to send commands over the SPI to the external Flash/SD. You only need to wait for the other open/write/... operation is finsihed.
I added a test that demonstrates that you may still have two files opened at the same time accessing them from different threads. Hope that helps.

File: doc/releases/release-notes-3.3.rst (hunk 11 above)
Comment (Ox11) [reply doc/releases/release-notes-3.3.rst:513]: Makes sense. However, do you know when 3.3 will be released and the 3.4 file created? - Doesn't this hold up all pending PRs?

File: modules/fatfs/zephyr_fatfs_config.h (hunk 12 above)
Comment (de-nordic) [reply modules/fatfs/zephyr_fatfs_config.h:75]: > I'm not familiar with the internals of how this works, but let's assume that there is one partition with a log file and any other random file, if the log file is opened at startup and let's say it moves to a new file once an hour (or even once a day) but there's a way for files to be retrieved from this file system - what happens if a client requests the "other random file", does their file access part of their code (let's assume MCUmgr) then halt for **1 hour/1 day** until the mutex can be taken? If so, that seems rather inconvenient for all users of this, and unacceptable

This will not be a problem. Generally most of functions, there are exceptions, call `validate` (https://github.com/zephyrproject-rtos/fatfs/blob/master/ff.c#L3613) that does locking for the operation, and do `LEAVE_FF` to retrun that does unlocking, https://github.com/zephyrproject-rtos/fatfs/blob/427159bf95ea49b7680facffaa29ad506b42709b/ff.c#L237.

File: subsys/fs/Kconfig.fatfs [hunk 13]
```diff
@@ -211,6 +212,13 @@ config FS_FATFS_WINDOW_ALIGNMENT
 	  that, in worst scenario, value provided here may cause FATFS
 	  structure to have size of twice the value.
 
+config FS_FATFS_REENTRANT
+	bool "FatFs reentrant"
+	help
```
Comment (arbrauns) [inline subsys/fs/Kconfig.fatfs:217]: Would a `depends on !FS_FATFS_LFN_MODE_BSS` make sense here? Without, it's really easy to enable LFN in its default configuration, enable reentrancy, and not realize that it's actually still not thread safe. With the `depends`, there would at least be a warning because `CONFIG_FS_FATFS_REENTRANT` can't be selected.

Comment (de-nordic) [reply subsys/fs/Kconfig.fatfs:217]: Yeah, we should probably add that.

Comment (Ox11) [reply subsys/fs/Kconfig.fatfs:217]: Good point!
I tested it locally with `depends on !FS_FATFS_LFN_MODE_BSS`:
I added the `CONFIG_FS_FATFS_LFN=y` in the testcase.yaml for the reentrant test in the `extra_configs`, but the test did not fail until I set the mode manually with `FS_FATFS_LFN_MODE_BSS=y`. I also tested it by adding the `depends on !FS_FATFS_REENTRANT` to `FS_FATFS_LFN_MODE_BSS` and had the same behaviour.
The default mode is probably the most critical choise to warn about which still won't work.
Do you know more about this behaviour?

Comment (Ox11) [reply subsys/fs/Kconfig.fatfs:218]: I added `depends on !FS_FATFS_REENTRANT` to `FS_FATFS_LFN_MODE_BSS` because `FS_FATFS_LFN_MODE_BSS` is not always visible and therefore I think this way round it is cleaner.

**Review by nordicjm — COMMENTED** [review]
this can rebased now to use the 3.4 release note file

File: MAINTAINERS.yml [hunk 14]
```diff
@@ -1240,6 +1240,16 @@ Filesystems:
   labels:
     - "area: File System"
 
+"Filesystems: FatFs":
```
[the remark this answers is not in the document]
Comment (Ox11) [reply MAINTAINERS.yml:1350]: Done.
But I am not quite sure if i should add me as a maintainer because I feel like I am just a collaborator. But I don't know if I can add a subsystem entry without a maintainer. Any ideas? @de-nordic @nordicjm

**Comment by Ox11** [comment]
> this can rebased now to use the 3.4 release note file

Done.

**Review by nordicjm — DISMISSED** [review]

File: MAINTAINERS.yml [hunk 15]
```diff
@@ -1347,6 +1347,16 @@ Filesystems:
...
```
Comment (de-nordic) [inline MAINTAINERS.yml:1350]: Rename this as "FatFs reentrant support"

File: MAINTAINERS.yml [hunk 16]
```diff
@@ -1347,6 +1347,16 @@ Filesystems:
...
+  status: maintained
+  maintainers:
+    - ox11
+  files:
+    - subsys/fs/fat_fs.c
```
Comment (de-nordic) [inline MAINTAINERS.yml:1355]: I see no reason to pass Maintainership on this file, it is not even touched in this PR. Remove it from list.

File: MAINTAINERS.yml [hunk 17]
```diff
@@ -1347,6 +1347,16 @@ Filesystems:
...
+    - modules/fatfs/zfs_ffsystem.c
```
Comment (de-nordic) [inline MAINTAINERS.yml:1356]: List is missing test_fat_file_reentrant.c file.

**Review by de-nordic — CHANGES_REQUESTED** [review]

File: MAINTAINERS.yml (hunk 15 above)
Comment (Ox11) [reply MAINTAINERS.yml]: Done.

File: MAINTAINERS.yml (hunk 16 above)
Comment (Ox11) [reply MAINTAINERS.yml]: Yes, you are right.

File: MAINTAINERS.yml (hunk 17 above)
Comment (Ox11) [reply MAINTAINERS.yml:1355]: Added it.

**Comment by Ox11** [comment]
Could you please review this PR again @nordicjm @de-nordic ? Did I miss something that I have to do?

**Review by nordicjm — APPROVED** [review]

**Review by de-nordic — APPROVED** [review]

**Comment by Ox11** [comment]
Thank you for your approves @nordicjm @de-nordic. Who can/will merge this PR? - I canot merge it, as I am not authorized.

File: MAINTAINERS.yml [hunk 18]
```diff
@@ -1371,6 +1371,16 @@ Filesystems:
...
 
+"Filesystems: FatFs reentrant support":
+  status: maintained
+  maintainers:
+    - ox11
+  files:
+    - modules/fatfs/zfs_ffsystem.c
+    - tests/subsys/fs/fat_fs_api/src/test_fat_file_reentrant.c
+  labels:
+    - "area: File System"
```
Comment (stephanosio) [inline MAINTAINERS.yml:1382]: [From the documentation:](https://docs.zephyrproject.org/latest/project/project_roles.html#maintainers-file)

> We should keep the granularity of code maintainership at a manageable level

It does not make sense to have "FatFs reentrant support" as a dedicated area -- this is too much granularity.

Please revert this change or define the maintainership as "Filesystems: Fatfs."

Comment (Ox11) [reply MAINTAINERS.yml:1382]: I am not a maintainer of the entire FatFS module.
My suggestion would be to either create a new module like this
```
"Filesystems: FatFs":
  status: maintained
  maintainers:
    - de-nordic
  collaborators:
    - ox11
  files:
    - modules/fatfs
    - tests/subsys/fs/common
    - tests/subsys/fs/fat_fs_api
    - tests/subsys/fs/fat_fs_dual_drive
  labels:
    - "area: File System"
```

or to add myself to the collaborators in the fs area
```
Filesystems:
  status: maintained
  maintainers:
    - de-nordic
  collaborators:
    - Laczen
    - nashif
    - ox11
  files:
    - include/zephyr/fs/
    - samples/subsys/fs/
    - subsys/fs/
    - tests/subsys/fs/
  labels:
    - "area: File System"
```

The first option should IMHO be created by @de-nordic as he would be mentioned as maintainer. Additionally, is there anyone else that should be mentioned as collaborator?
